pkg/controller/extensions/extension/extension.go

```diff
@@ -0,0 +1,251 @@
+	"github.com/crossplaneio/crossplane/pkg/logging"
+)
+
+const (
+	controllerName = "extension.extensions.crossplane.io"
+
+	reconcileTimeout      = 1 * time.Minute
+	requeueAfterOnSuccess = 10 * time.Second
+
+	reasonCreatingRBAC       = "failed to create RBAC"
+	reasonCreatingDeployment = "failed to create deployment"
+)
+
+var (
+	log              = logging.Logger.WithName(controllerName)
+	resultRequeue    = reconcile.Result{Requeue: true}
+	requeueOnSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
+)
+
+// Reconciler reconciles a Instance object
+type Reconciler struct {
+	kube client.Client
+	factory
+}
+
+// Add creates a new Extension Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	r := &Reconciler{
+		kube:    mgr.GetClient(),
+		factory: &extensionHandlerFactory{},
+	}
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Extension
+	err = c.Watch(&source.Kind{Type: &v1alpha1.Extension{}}, &controllerHandler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the Extension for a Instance object and makes changes based on the state read
+// and what is in the Instance.Spec
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("reconciling", "kind", v1alpha1.ExtensionKindAPIVersion, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	// fetch the CRD instance
+	i := &v1alpha1.Extension{}
+	if err := r.kube.Get(ctx, req.NamespacedName, i); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	handler := r.factory.newHandler(ctx, i, r.kube)
+
+	return handler.sync(ctx)
+}
+
+type handler interface {
+	sync(context.Context) (reconcile.Result, error)
+	create(context.Context) (reconcile.Result, error)
+	update(context.Context) (reconcile.Result, error)
+}
+
+type extensionHandler struct {
+	kube client.Client
+	i    *v1alpha1.Extension
+}
+
+type factory interface {
+	newHandler(context.Context, *v1alpha1.Extension, client.Client) handler
+}
+
+type extensionHandlerFactory struct{}
+
+func (f *extensionHandlerFactory) newHandler(ctx context.Context, i *v1alpha1.Extension, kube client.Client) handler {
+	return &extensionHandler{
+		kube: kube,
+		i:    i,
+	}
+}
+
+// ************************************************************************************************
+// Syncing/Creating functions
+// ************************************************************************************************
+func (h *extensionHandler) sync(ctx context.Context) (reconcile.Result, error) {
+	if h.i.Status.ControllerRef == nil {
```

Comment:
I think we should not infer the state of our application from the `status` field, instead, we should Observe the "system" in terms of check - is there an application (Deployment). 

From: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
> The status summarizes the current state of the object in the system, and is usually persisted with the object by an automated processes but may be generated on the fly. At some cost and perhaps some temporary degradation in behavior, the status could be reconstructed by observation if it were lost.

I.E., if status sub-resource was "accedentally" whipped out, the controller should be able to reconstruct it to some extent using observability of the system. I know, this is not a new issue and this is how we've done things in the crossplane in the past. However, I think in this case we can (and should) follow this paradigm. 

To be clear, I think it is totally fine to track children references. However, I don't think we should use them (anymore) to infer our application state and hence the execution flow.

Reply:
i'll capture this to follow up on